src/Libraries/CoreNodes/List.cs

```diff
@@ -525,6 +525,170 @@ public static IDictionary GroupByKey(IList list, IList keys)
+                        idxResult.Add(Enumerable.Range(start, currentGroup.Count).ToList());
+                        currentGroup = new List<object> { list[i] };
+                        start = i;
+                    }
+                }
+                result.Add(currentGroup);
+                idxResult.Add(Enumerable.Range(start, currentGroup.Count).ToList());
+            }
+            else
+            {
+                var groupedItems = new List<List<object>>();
+                var itemGroups = new List<object>();
+                var indexGroups = new List<List<object>>();
+
+                for (int i = 0; i < list.Count; i++)
+                {
+                    var item = list[i];
+                    bool added = false;
+                    for (int j = 0; j < itemGroups.Count; j++)
+                    {
+                        if (AreItemsSimilar(itemGroups[j], item, tolerance))
+                        {
+                            groupedItems[j].Add(item);
+                            indexGroups[j].Add(i);
+                            added = true;
+                            break;
+                        }
+                    }
+
+                    if (!added)
+                    {
+                        itemGroups.Add(item);
+                        groupedItems.Add(new List<object> { item });
+                        indexGroups.Add(new List<object> { i });
+                    }
+                }
+
+                foreach (var group in groupedItems)
+                {
+                    result.Add(group);
+                }
+
+                foreach (var group in indexGroups)
+                {
+                    idxResult.Add(group);
+                }
+            }
+
+            return new Dictionary<string, object>
+            {
+                { "groupedValues", result },
+                { "groupedIndices", idxResult }
+            };
+        }
+
+        private static bool AreItemsSimilar(object item1, object item2, double tolerance)
+        {
+            if (item1 is string str1 && item2 is string str2)
+            {
+                tolerance = Math.Round(tolerance);
+                if (tolerance > 5)
+                    throw new ArgumentException("Tolerance value for string comparison cannot be greater than 5.", nameof(tolerance));
+                return LevenshteinDistance(str1, str2) <= tolerance;
+            }
+            else if (item1 is IComparable comp1 && item2 is IComparable comp2)
+            {
+                try
+                {
+                    return Math.Abs(Convert.ToDouble(comp1) - Convert.ToDouble(comp2)) <= tolerance;
+                }
+                catch
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                return Object.Equals(item1, item2);
+            }
+        }
+
+        private static int LevenshteinDistance(string a, string b)
```

Comment:
Is this algorithm taken from some public source? Can we include a link here so we have something to refer to for future devs?

Reply:
Good idea, added source